yarn-project/sequencer-client/src/simulator/public_executor.ts

```diff
@@ -178,18 +183,36 @@ export class WorldStatePublicDB implements PublicStateDB {
    * @returns Nothing.
    */
   commit(): Promise<void> {
-    for (const [k, v] of this.uncommitedWriteCache) {
-      this.commitedWriteCache.set(k, v);
+    for (const [k, v] of this.checkpointedWriteCache) {
+      this.committedWriteCache.set(k, v);
+    }
+    // uncommitted writes take precedence over checkpointed writes
+    // since they are the most recent
+    for (const [k, v] of this.uncommittedWriteCache) {
+      this.committedWriteCache.set(k, v);
     }
-    return this.rollback();
+    return this.rollbackToCommit();
   }
 
   /**
    * Rollback the pending changes.
    * @returns Nothing.
    */
-  rollback(): Promise<void> {
-    this.uncommitedWriteCache = new Map<bigint, Fr>();
+  async rollbackToCommit(): Promise<void> {
+    await this.rollbackToCheckpoint();
+    this.checkpointedWriteCache = new Map<bigint, Fr>();
+    return Promise.resolve();
+  }
+
+  checkpoint(): Promise<void> {
+    for (const [k, v] of this.uncommittedWriteCache) {
+      this.checkpointedWriteCache.set(k, v);
+    }
+    return this.rollbackToCheckpoint();
+  }
```

Comment:
I think this over accumulates state if my tx has >1 functions in app logic :thinking::
1. my setup fn runs correctly, `checkpoint()`
2. my first app logic fn runs correctly, `checkopint()`
3. my second app logic fn reverts, `rollBackToCheckpoint()` <-- I think this only undoes the second fn's state changes when it should rollback to setup?

Reply:
Actually, it does not run checkpoint until the end of each phase. That is, `handle` is only called once per phase, and we checkpoint at the end of it.